server/src/main/java/org/opensearch/rest/NamedRoute.java

```diff
@@ -9,47 +9,140 @@
 package org.opensearch.rest;
 
 import org.opensearch.OpenSearchException;
+import org.opensearch.transport.TransportService;
+
+import java.util.HashSet;
+import java.util.Set;
 
 /**
  * A named Route
  *
  * @opensearch.internal
  */
 public class NamedRoute extends RestHandler.Route {
+
     private static final String VALID_ACTION_NAME_PATTERN = "^[a-zA-Z0-9:/*_]*$";
     static final int MAX_LENGTH_OF_ACTION_NAME = 250;
 
-    private final String name;
+    private final String uniqueName;
+    private final Set<String> actionNames;
 
-    public boolean isValidRouteName(String routeName) {
-        if (routeName == null || routeName.isBlank() || routeName.length() > MAX_LENGTH_OF_ACTION_NAME) {
-            return false;
+    /**
+     * Builder class for constructing instances of {@link NamedRoute}.
+     */
+    public static class Builder {
+        private RestRequest.Method method;
+        private String path;
+        private String uniqueName;
+        private Set<String> legacyActionNames;
+
+        /**
+         * Sets the REST method for the route.
+         *
+         * @param method the REST method for the route
+         * @return the builder instance
+         */
+        public Builder method(RestRequest.Method method) {
+            this.method = method;
+            return this;
+        }
+
+        /**
+         * Sets the URL path for the route.
+         *
+         * @param path the URL path for the route
+         * @return the builder instance
+         */
+        public Builder path(String path) {
+            this.path = path;
+            return this;
+        }
+
+        /**
+         * Sets the name for the route.
+         *
+         * @param name the name for the route
+         * @return the builder instance
+         */
+        public Builder uniqueName(String name) {
+            this.uniqueName = name;
+            return this;
+        }
+
+        /**
+         * Sets the legacy action names for the route.
+         *
+         * @param legacyActionNames the legacy action names for the route
+         * @return the builder instance
+         */
+        public Builder legacyActionNames(Set<String> legacyActionNames) {
+            this.legacyActionNames = legacyActionNames;
+            return this;
```

Comment:
```suggestion
        public Builder legacyActionNames(Set<String> legacyActionNames) {
            this.legacyActionNames.addAll(legacyActionNames);
            return this;
```
Don't allow modification of the underlying set

Reply:
There is no underlying instance till it reaches this point as legacyActionNames set is not instantiated.